### PR TITLE
✨CEPH/S3 additions (⚠️ devops)

### PR DESCRIPTION
**Pull request by GitHK**
# ⚠️ devops

The filestash services was added to the osparc ops stack, some scripts might require porting to the ops repo.

## What do these changes do?

- Adds a new service to browse/edit/upload/delete/visualise S3 files.
- The service is exposed on poet 9002 next to 9001 (used by minio/flower).
- To use it you need to input the access_key and the secret_key just as you did for minio.
- Adds settings option to select rclone's `vfs-cache-mode`

### CODE REVIEW

File: scripts/filestash/create_config.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+"""
+
+import os
+
+from pathlib import Path
+from string import Template
+
+SCRIPT_DIR = Path(__file__).absolute().parent
+TEMPLATE_PATH = SCRIPT_DIR / "filestash_config.json.template"
+CONFIG_TEMP_DIR = SCRIPT_DIR / ".." / ".." / "_tmp"
```
Comment (pcrespov) [inline scripts/filestash/create_config.py:15]: TIP: 
```python
script_dir = Path(__file__).resolve().parent
```

**Comment by KZzizzle** [comment]
thanks for the notice!

File: scripts/filestash/create_config.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def main() -> None:
+    patch_env_vars()
+
+    assert TEMPLATE_PATH.exists()
+
+    template_content = TEMPLATE_PATH.read_text()
+
+    config_json = Template(template_content).substitute(os.environ)
```
Comment (pcrespov) [inline scripts/filestash/create_config.py:37]: nice, i never used this one before :-)

File: scripts/filestash/create_config.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    CONFIG_TEMP_DIR.mkdir(parents=True, exist_ok=True)
```
Comment (pcrespov) [inline scripts/filestash/create_config.py:39]: temp files can be tricky. YOu cannot guarantee that whoever runs this script has e.g. permissions to create this directory where you defined it.
For temp directories i hightly recommend you ``tempfile``. It guarantees the creation and deletion of the tmpdir

https://docs.python.org/3/library/tempfile.html#examples

File: .gitignore [hunk 4]
```diff
@@ -171,3 +171,5 @@ packages/service-integration/compose-spec.json
 
 # temporary files by vscode
 Untitled*
+
+_tmp/
```
Comment (pcrespov) [inline .gitignore:175]: not a good idea. Check comments about tempdir below

Comment (pcrespov) [reply .gitignore:175]: NOTE2: when you want to create an artifact and want to ignore it, just add in the name the keyword ``ignore`` 
for example ``myfile.txt`` -> ``myfile.ignore.txt``

File: scripts/filestash/create_config.py (hunk 1 above)
Comment (GitHK) [reply scripts/filestash/create_config.py:15]: nice one, actually resolves it cleanly

File: scripts/filestash/filestash_config.json.template [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  },
+  "email": {
+      "server": null,
+      "port": null,
+      "username": null,
+      "password": null,
+      "from": null
+  },
+  "auth": {
+      "admin": "$$2a$$10$$viR16hXd35bAaEJFEgpd9OqIzNBb/VoIsgQ8P3SjKxolpEQEHltrW"
```
Comment (mrnicegyu11) [inline scripts/filestash/filestash_config.json.template:74]: How is this generated? I'd like to auto-generated in the ops-stack

Comment (mrnicegyu11) [reply scripts/filestash/filestash_config.json.template:74]: @GitHK

File: scripts/filestash/filestash_config.json.template [hunk 6]
```diff
@@ -0,0 +1,90 @@
+{
+  "general": {
+      "name": null,
+      "port": null,
+      "host": null,
+      "secret_key": "zUMyrLIDRMJKEdWw",
```
Comment (mrnicegyu11) [inline scripts/filestash/filestash_config.json.template:6]: How is this generated, can we update this with our common secret? @GitHK

Comment (GitHK) [reply scripts/filestash/filestash_config.json.template:6]: It is just some random string. I can randomise within this script and replace .

File: scripts/filestash/filestash_config.json.template (hunk 5 above)
Comment (GitHK) [reply scripts/filestash/filestash_config.json.template:74]: Looks just like this https://docs.docker.com/registry/deploying/#native-basic-auth

File: scripts/filestash/filestash_config.json.template (hunk 6 above)
Comment (GitHK) [reply scripts/filestash/filestash_config.json.template:6]: OK, I've added this

File: scripts/filestash/filestash_config.json.template (hunk 5 above)
Comment (GitHK) [reply scripts/filestash/filestash_config.json.template:74]: I'm sorry it has nothing to do with this. It was generated by the program.
When you first start it it will ask for admin credentials. I've only escaped the `$` signs in there.

**Review by pcrespov — APPROVED** [review]

File: services/docker-compose-ops.yml [hunk 7]
```diff
@@ -69,6 +69,24 @@ services:
       timeout: 20s
       retries: 3
 
+  filestash:
+    container_name: filestash
+    image: machines/filestash:3a01b70
+    restart: always
+    ports:
+      - "9002:8334"
```
Comment (mguidon) [inline services/docker-compose-ops.yml:77]: is that a good port? I think minio 1-4 run on 9000-9003.

Comment (GitHK) [reply services/docker-compose-ops.yml:77]: This is the local published port. Usually minio is exposed on 9001. Not sure if this will collide with something else.

**Review by mguidon — APPROVED** [review]

File: packages/pytest-simcore/src/pytest_simcore/docker_compose.py [hunk 8]
```diff
@@ -154,9 +156,46 @@ def simcore_docker_compose(
...
+    create_filestash_config_py = (
+        osparc_simcore_scripts_dir / "filestash" / "create_config.py"
+    )
+
+    # ensures .env at git_root_dir, which will be used as current directory
+    assert env_file_for_testing.exists()
+    env_values = dotenv_values(env_file_for_testing)
+
+    process = subprocess.run(
+        ["python3", f"{create_filestash_config_py}"],
```
Comment (pcrespov) [inline packages/pytest-simcore/src/pytest_simcore/docker_compose.py:174]: TIP: why not taking advantage that this is python and you just import it to get the path?

**Review by pcrespov — APPROVED** [review]

File: Makefile [hunk 9]
```diff
@@ -77,8 +77,8 @@ endif
 get_my_ip := $(shell hostname --all-ip-addresses | cut --delimiter=" " --fields=1)
 
 # NOTE: this is only for WSL2 as the WSL2 subsystem IP is changing on each reboot
-S3_ENDPOINT := $(get_my_ip):9001
-export S3_ENDPOINT
+#S3_ENDPOINT := $(get_my_ip):9001
+#export S3_ENDPOINT
```
Comment (sanderegg) [inline Makefile:81]: could you please remove the lines?

File: Makefile [hunk 10]
```diff
@@ -195,16 +195,27 @@ CPU_COUNT = $(shell cat /proc/cpuinfo | grep processor | wc -l )
...
+	# Compiling config file for filestash
+	$(eval TMP_PATH_TO_FILESTASH_CONFIG=$(shell set -o allexport; \
+	source $(CURDIR)/.env; \
+	set +o allexport; \
+	python3 scripts/filestash/create_config.py))
 	# Creating config for ops stack to $@
-	@docker-compose --env-file .env --file services/docker-compose-ops.yml --log-level=ERROR config > $@
+	# -> filestash config at $(TMP_PATH_TO_FILESTASH_CONFIG)
+	@$(shell \
+		export TMP_PATH_TO_FILESTASH_CONFIG="${TMP_PATH_TO_FILESTASH_CONFIG}"; \
```
Comment (sanderegg) [inline Makefile:206]: use ```&&``` instead of ```;``` it's a better practice. else if one fail it will not be discovered.

File: scripts/filestash/create_config.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
+"""
+Parses the configuration template and injects it where the platform expects it
+Notes:
+- Admin credentials for filestash are admin:adminadmin
+- $ must be escaped with $$ in the teemplate file
```
Comment (sanderegg) [inline scripts/filestash/create_config.py:5]:
```suggestion
- $ must be escaped with $$ in the template file
```

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/_namepsace.py [hunk 12]
```diff
@@ -0,0 +1,10 @@
+from models_library.projects_nodes import NodeID
+
+from ...models.schemas.constants import DYNAMIC_SIDECAR_SERVICE_PREFIX
+
+
+def get_compose_namespace(node_uuid: NodeID) -> str:
+    # To avoid collisions for started docker resources a unique identifier is computed:
+    # - avoids container level collisions on same node
+    # - avoids volume level collisions on same node
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/_namepsace.py:9]: what about multiple deployments on same swarm? if a study would be copied from staging to prod, then the node_uuid would be the same in both

File: services/docker-compose-ops.yml [hunk 13]
```diff
@@ -69,6 +69,24 @@ services:
...
+    volumes:
+      - ${TMP_PATH_TO_FILESTASH_CONFIG}:/app/data/state/config/config.json
+    networks:
+      - simcore_default
+
+  onlyoffice:
```
Comment (sanderegg) [inline services/docker-compose-ops.yml:83]: what is this?

File: services/web/server/tests/unit/with_dbs/10/test_resource_manager.py [hunk 14]
```diff
@@ -414,7 +414,7 @@ async def test_interactive_services_removed_after_logout(
         reraise=True, stop=stop_after_delay(10), wait=wait_fixed(1)
     ):
         with attempt:
-            mocked_director_v2_api["director_v2_core.stop_service"].assert_awaited_with(
+            mocked_director_v2_api["director_v2_core.stop_service"].assert_any_await(
```
Comment (sanderegg) [inline services/web/server/tests/unit/with_dbs/10/test_resource_manager.py:417]: are you sure about this change?

**Review by sanderegg — APPROVED** [review]

File: Makefile (hunk 9 above)
Comment (GitHK) [reply Makefile:81]: I really though I did remove them!

File: services/docker-compose-ops.yml (hunk 13 above)
Comment (GitHK) [reply services/docker-compose-ops.yml:83]: It's a service used to preview documents inside filestash. It also allows you to edit some types of files and save them back.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/_namepsace.py (hunk 12 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/_namepsace.py:9]: Well if done via the import feature, it changes the project_id and node_ids in the study. This should not be an issue.

I could append the `SWARM_STACK_NAME`, It should work like that as well.

File: services/web/server/tests/unit/with_dbs/10/test_resource_manager.py (hunk 14 above)
Comment (GitHK) [reply services/web/server/tests/unit/with_dbs/10/test_resource_manager.py]: Let me revert it and check what is happening there if it still fails. In theory it should be equivalent, but there might be and edge case here.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/_namepsace.py (hunk 12 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/_namepsace.py:9]: This apparently breaks more that one thing. I will create an issue for it.
